src/tiledb/cloud/vcf/query.py

```diff
@@ -0,0 +1,308 @@
+    logger.debug("Max memory usage: %0.3f GiB", memory_usage_gb)
+    logger.debug("Incomplete queries: %d", len(tables) - 1)
+    logger.debug("Records read: %d", table.num_rows)
+    logger.debug("Arrow table size: %0.3f MiB", table.nbytes / (1 << 20))
+
+    return table
+
+
+def concat_tables_udf(
+    tables: List[pa.Table],
+    *,
+    config: Optional[Mapping[str, Any]] = None,
+    log_uri: Optional[str] = None,
+) -> pa.table:
+    """
+    Concatenate a list of Arrow tables.
+
+    :param tables: Arrow tables
+    :param config: config dictionary, defaults to None
+    :param log_uri: log URI for profiling, defaults to None
+    :return: concatenated Arrow table
+    """
+
+    logger = setup(config)
+
+    with Profiler(array_uri=log_uri) as prof:
+        table = pa.concat_tables(
+            [x for x in tables if x is not None and x.num_rows > 0]
+        )
+        prof.write("result", table.num_rows, table.nbytes)
+
+    memory_usage_gb = max_memory_usage() / (1 << 30)
+    logger.debug("Max memory usage: %0.3f GiB", memory_usage_gb)
+    logger.debug("Records read: %d", table.num_rows)
+    logger.debug("Arrow table size: %0.3f MiB", table.nbytes / (1 << 20))
+
+    return table
+
+
+# --------------------------------------------------------------------
+# DAGs
+# --------------------------------------------------------------------
+
+
+def read(
+    dataset_uri: str,
+    *,
+    config: Optional[Mapping[str, Any]] = None,
+    attrs: Optional[Union[Sequence[str], str]] = None,
+    regions: Optional[Union[Sequence[str], str]] = None,
+    bed_file: Optional[str] = None,
+    num_region_partitions: int = 1,
+    max_workers: int = MAX_WORKERS,
+    samples: Optional[Union[Sequence[str], str]] = None,
+    memory_budget_mb: int = 1024,
+    af_filter: Optional[str] = None,
+    log_uri: Optional[str] = None,
+    namespace: Optional[str] = None,
+    resource_class: Optional[str] = None,
+    return_dag: bool = False,
+) -> Union[pa.Table, Tuple[tiledb.cloud.dag.DAG, tiledb.cloud.dag.Node]]:
+    """
+    Run a distributed read on a TileDB-VCF dataset.
+
+    :param dataset_uri: dataset URI
+    :param config: config dictionary, defaults to None
+    :param attrs: attribute names to read, defaults to None
+    :param regions: genomics regions to read, defaults to None
+    :param bed_file: URI of a BED file containing genomics regions to read, defaults to None
+    :param num_region_partitions: number of region partitions, defaults to 1
+    :param samples: sample names to read, defaults to None
+    :param memory_budget_mb: VCF memory budget in MiB, defaults to 1024
+    :param af_filter: allele frequency filter, defaults to None
+    :param log_uri: log array URI for profiling, defaults to None
+    :param namespace: TileDB-Cloud namespace, defaults to None
+    :param resource_class: TileDB-Cloud resource class for UDFs, defaults to None
+    :param return_dag: return a DAG instead of the query results, defaults to False
```

Comment:
I'm not a fan of "return one thing, or if this parameter is set, a completely different thing". Maybe make two functions, one which forwards its arguments to the other and returns the output?

```python
def build_read_dag(...) -> Tuple[...]:
  ...

def read(dataset_uri: str, **kwargs) -> ...:
  dag, table = build_read_dag(dataset_uri, **kwargs)
  # ...
  return ...
```

Reply:
Ok, I'll make that change.

Reply:
I ended up keeping the full function signature for `read` to preserve the docstrings, since these are user facing functions.